src/user-meta/domain/custom-meta-interface.php

```diff
@@ -27,4 +34,13 @@ public function get_field_id(): string;
 	 * @return bool
 	 */
 	public function is_empty_allowed(): bool;
+
+	/**
+	 * Renders the custom meta's field in the user form.
+	 *
+	 * @param int $user_id The user ID.
+	 *
+	 * @return void
+	 */
+	public function render_field( $user_id ): void;
```

Comment:
I agree that it's not the most puristic way to deal with this, but I would insist on doing so. My thoughts in bullet points:
* The current abstraction of the rendering is as far as we could go,without getting into over-engineering territory, since the rendering of each meta is similar but not identical.
* The current approach is not adhering to strict 🧅/DDD rules, but it's close enough for me.
* One even better solution would be to offload this to JS land, like you said. But, that seems like an overkill for me that would complicate such a small task.
  * And if we ever want to do a JS solution, we are νος way better suited after this PR, now that the data are much more contained and structured. So, all this work won't go to waste.

All in all I think that this is good enough and also paves the way for even better handling in the future

Reply:
I guess that is fair enough.